### PR TITLE
Changed erase to just noop if erases are too small for erase command

### PR DESCRIPTION
**Pull request by geky**
Since it's an optimization, we should only send the command if the range of bytes being erased fits. get_erase_size should still report 512, and the erase should just noop with <SECTOR_SIZE bytes being erased. If filesystems erase a larger range (at least spiffs does this), the optimization can kick in.

see https://github.com/ARMmbed/mbed-os/issues/4899#issuecomment-323092120
cc @deepikabhavnani, @LemonBoy

### CODE REVIEW

File: SDBlockDevice.cpp [hunk 1]
```diff
@@ -558,6 +558,17 @@ int SDBlockDevice::erase(bd_addr_t addr, bd_size_t size)
     }
     int status = BD_ERROR_OK;
 
+    // Align erases, missed blocks will be left unerased, but this is fine as
+    // the block device api doesn't garuntee the final state of erased bytes and
+    // a following program will still work fine.
+    bd_size_t misalignment = _erase_size - (addr % _erase_size);
```
Comment (LemonBoy) [inline SDBlockDevice.cpp]: Looks good, I'd go for something simpler such as
```c
if (size < _erase_size)
    return status; // it's probably wise to inform the user the operation is unsupported
```

as the controller already rounds the start/end addresses according to the required granularity.

Comment (deepikabhavnani) [inline SDBlockDevice.cpp]: @geky Agreed with @LemonBoy. SD controller will take care of alignment and deciding what range of erase blocks should be erased in particular request.

We can at most throw an error, if size and address are not aligned to erase size, which should be done by is_valid_erase() check.

Comment (geky) [inline SDBlockDevice.cpp]: Ah ok, should be updated

**Comment by geky** [comment]
@deepikabhavnani, This patch also changes get_erase_size to return the _block_size. If _block_size is smaller than _erase_size, unaligned "erases" are possible.

**Comment by deepikabhavnani** [comment]
@geky You are returning success from erase, from filesystem point of view data is erased and will try to overwrite. Might have issues when the card gets full to certain limit. 
And if you update the code to send error than it is equivalent to existing one.

File: SDBlockDevice.cpp [hunk 2]
```diff
@@ -558,6 +558,14 @@ int SDBlockDevice::erase(bd_addr_t addr, bd_size_t size)
...
 
+    // Toss out erases that are too small, this is fine as the block device api
+    // doesn't garuntee the final state of erased bytes and a following program
+    // will still work fine.
+    if (size < _erase_size) {
+        _lock.unlock();
+        return status;
+    }
+
     size -= _block_size;
```
Comment (LemonBoy) [inline SDBlockDevice.cpp]: Should this be `_erase_size`?

Comment (deepikabhavnani) [inline SDBlockDevice.cpp]: Yes it should be _erase_size.

Comment (geky) [inline SDBlockDevice.cpp]: Ah ok 👍

File: SDBlockDevice.cpp [hunk 3]
```diff
@@ -558,7 +558,15 @@ int SDBlockDevice::erase(bd_addr_t addr, bd_size_t size)
...
-    size -= _block_size;
+    // Toss out erases that are too small, this is fine as the block device api
+    // doesn't garuntee the final state of erased bytes and a following program
+    // will still work fine.
+    if (size < _erase_size) {
+        _lock.unlock();
+        return status;
+    }
+
+    size -= _erase_size;
```
Comment (deepikabhavnani) [inline SDBlockDevice.cpp]: This should be block_size. SD card needs start and end block address to perform erase, hence size of block(last) is deducted to calculate end block address properly.

File: SDBlockDevice.cpp [hunk 4]
```diff
@@ -594,7 +602,7 @@ bd_size_t SDBlockDevice::get_program_size() const
 
 bd_size_t SDBlockDevice::get_erase_size() const
 {
-    return _erase_size;
+    return _block_size;
```
Comment (deepikabhavnani) [inline SDBlockDevice.cpp]: Can we leave a comment, stating erase size is used for erase operations of flash devices in FAT filesystem and should be equivalent to block size, as no erase is required for 
sd cards before every write operation.

Comment (geky) [reply SDBlockDevice.cpp]: I would actually propose we remove the erase function from this driver completely. See https://github.com/ARMmbed/mbed-os/pull/4983. Updating...

Comment (deepikabhavnani) [reply SDBlockDevice.cpp]: @geky - Work in progress, will remove erase function and add trim instead

**Pull request merged**